File: packages/core/src/programmers/http/HttpAssertFormDataProgrammer.ts
import ts from "typescript";

import { IProgrammerProps } from "../../context/IProgrammerProps";
import { ITypiaContext } from "../../context/ITypiaContext";
import { IdentifierFactory } from "../../factories/IdentifierFactory";
import { StatementFactory } from "../../factories/StatementFactory";
import { TypeFactory } from "../../factories/TypeFactory";
import { AssertProgrammer } from "../AssertProgrammer";
import { FunctionProgrammer } from "../helpers/FunctionProgrammer";
import { FeatureProgrammer } from "../internal/FeatureProgrammer";
import { HttpFormDataProgrammer } from "./HttpFormDataProgrammer";

export namespace HttpAssertFormDataProgrammer {
  export const decompose = (props: {
    context: ITypiaContext;
    functor: FunctionProgrammer;
    type: ts.Type;
    name: string | undefined;
    init?: ts.Expression | undefined;
  }): FeatureProgrammer.IDecomposed => {
    const assert: FeatureProgrammer.IDecomposed = AssertProgrammer.decompose({
      ...props,
      context: {
        ...props.context,
        options: {
          ...props.context.options,
          functional: false,
          numeric: true,
        },
      },
      config: {
        equals: false,
        guard: false,
      },
    });
    const decode: FeatureProgrammer.IDecomposed =
      HttpFormDataProgrammer.decompose(props);
    return {
      functions: {
        ...assert.functions,
        ...decode.functions,
      },
      statements: [
        ...assert.statements,
        ...decode.statements,
        StatementFactory.constant({
          name: "__assert",
          value: assert.arrow,
        }),
        StatementFactory.constant({
          name: "__decode",
          value: decode.arrow,
        }),
      ],
      arrow: ts.factory.createArrowFunction(
        undefined,
        undefined,
        [
          IdentifierFactory.parameter("input", TypeFactory.keyword("any")),
          AssertProgrammer.Guardian.parameter({
            context: props.context,
            init: props.init,
          }),
        ],
        decode.arrow.type,
        undefined,
        ts.factory.createCallExpression(
          ts.factory.createIdentifier("__assert"),
          undefined,
          [
            ts.factory.createCallExpression(
              ts.factory.createIdentifier("__decode"),
              undefined,
              [ts.factory.createIdentifier("input")],
            ),
            AssertProgrammer.Guardian.identifier(),
          ],
        ),
      ),
    };
  };

  export const write = (props: IProgrammerProps): ts.CallExpression => {
    const functor: FunctionProgrammer = new FunctionProgrammer(
      props.modulo.getText(),
    );
    const result: FeatureProgrammer.IDecomposed = decompose({
      ...props,
      functor,
    });
    return FeatureProgrammer.writeDecomposed({
      modulo: props.modulo,
      functor,
      result,
    });
  };
}

File: packages/typia/native/core/programmers/http/HttpAssertFormDataProgrammer.go
package http

import (
	shimast "github.com/microsoft/typescript-go/shim/ast"
	shimchecker "github.com/microsoft/typescript-go/shim/checker"
	nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
	nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
	nativeprogrammers "github.com/samchon/typia/packages/typia/native/core/programmers"
	nativehelpers "github.com/samchon/typia/packages/typia/native/core/programmers/helpers"
	nativeinternal "github.com/samchon/typia/packages/typia/native/core/programmers/internal"
)

type httpAssertFormDataProgrammerNamespace struct{}

var HttpAssertFormDataProgrammer = httpAssertFormDataProgrammerNamespace{}

type HttpAssertFormDataProgrammer_DecomposeProps struct {
	Context nativecontext.ITypiaContext
	Functor *nativehelpers.FunctionProgrammer
	Type    *shimchecker.Type
	Name    *string
	Init    *shimast.Node
}

var httpAssertProgrammer_factory = shimast.NewNodeFactory(shimast.NodeFactoryHooks{})

func (httpAssertFormDataProgrammerNamespace) Decompose(props HttpAssertFormDataProgrammer_DecomposeProps) nativeinternal.FeatureProgrammer_IDecomposed {
	assert := nativeprogrammers.AssertProgrammer.Decompose(nativeprogrammers.AssertProgrammer_DecomposeProps{
		Context: httpProgrammer_context(props.Context, false, true),
		Functor: props.Functor,
		Config: nativeprogrammers.AssertProgrammer_IConfig{
			Equals: false,
			Guard:  false,
		},
		Type: props.Type,
		Name: props.Name,
		Init: props.Init,
	})
	decode := HttpFormDataProgrammer.Decompose(HttpFormDataProgrammer_DecomposeProps{
		Context: props.Context,
		Functor: props.Functor,
		Type:    props.Type,
		Name:    props.Name,
	})
	return httpProgrammer_assert_result(props.Context, props.Init, assert, decode)
}

func (httpAssertFormDataProgrammerNamespace) Write(props nativecontext.IProgrammerProps) *shimast.Node {
	functor := nativehelpers.NewFunctionProgrammer(httpProgrammer_method_text(props.Modulo))
	result := HttpAssertFormDataProgrammer.Decompose(HttpAssertFormDataProgrammer_DecomposeProps{
		Context: props.Context,
		Functor: functor,
		Type:    props.Type,
		Name:    props.Name,
		Init:    props.Init,
	})
	return nativeinternal.FeatureProgrammer.WriteDecomposed(nativeinternal.FeatureProgrammer_WriteDecomposedProps{
		Modulo:  props.Modulo,
		Functor: functor,
		Result:  result,
	})
}

func httpProgrammer_assert_result(context nativecontext.ITypiaContext, init *shimast.Node, assert nativeinternal.FeatureProgrammer_IDecomposed, decode nativeinternal.FeatureProgrammer_IDecomposed) nativeinternal.FeatureProgrammer_IDecomposed {
	decodeArrow := decode.Arrow.AsArrowFunction()
	statements := append([]*shimast.Node{}, assert.Statements...)
	statements = append(statements, decode.Statements...)
	statements = append(statements,
		nativefactories.StatementFactory.Constant(nativefactories.StatementFactory_ConstantProps{
			Name:  "__assert",
			Value: assert.Arrow,
		}),
		nativefactories.StatementFactory.Constant(nativefactories.StatementFactory_ConstantProps{
			Name:  "__decode",
			Value: decode.Arrow,
		}),
	)
	return nativeinternal.FeatureProgrammer_IDecomposed{
		Functions:  httpProgrammer_merge_functions(assert.Functions, decode.Functions),
		Statements: statements,
		Arrow: httpAssertProgrammer_factory.NewArrowFunction(
			nil,
			nil,
			httpAssertProgrammer_factory.NewNodeList([]*shimast.Node{
				nativefactories.IdentifierFactory.Parameter("input", nativefactories.TypeFactory.Keyword("any"), nil),
				nativeprogrammers.Guardian.Parameter(struct {
					Context nativecontext.ITypiaContext
					Init    *shimast.Node
				}{
					Context: context,
					Init:    init,
				}),
			}),
			decodeArrow.Type,
			nil,
			httpAssertProgrammer_factory.NewToken(shimast.KindEqualsGreaterThanToken),
			httpAssertProgrammer_factory.NewCallExpression(
				httpAssertProgrammer_factory.NewIdentifier("__assert"),
				nil,
				nil,
				httpAssertProgrammer_factory.NewNodeList([]*shimast.Node{
					httpAssertProgrammer_factory.NewCallExpression(
						httpAssertProgrammer_factory.NewIdentifier("__decode"),
						nil,
						nil,
						httpAssertProgrammer_factory.NewNodeList([]*shimast.Node{httpAssertProgrammer_factory.NewIdentifier("input")}),
						shimast.NodeFlagsNone,
					),
					nativeprogrammers.Guardian.Identifier(),
				}),
				shimast.NodeFlagsNone,
			),
		),
	}
}
